File: .archive/lib-ts/history.ts
/**
 * Audit log reconstruction from Automerge's built-in change history.
 *
 * Instead of maintaining a separate auditLog array on the CRDT document,
 * we embed structured metadata in each Automerge change's `message` field
 * (see operations.ts) and reconstruct the audit log at read time using
 * Automerge's getChangesMetaSince() API.
 *
 * This eliminates ~72% of document size overhead from the old approach
 * and avoids the paradox where clearing/trimming the audit log actually
 * increases the CRDT history size.
 */

import * as Automerge from "#automerge";
import type { Project, AuditEntry } from "./schema.js";
import type { ChangeMessage } from "./operations.js";

type Doc = Automerge.Doc<Project>;

interface ChangeMeta {
  actor: string;
  seq: number;
  time: number;
  message: string | null;
  hash: string;
  deps: string[];
}

/**
 * Reconstruct the audit log from Automerge's change history.
 *
 * Uses getChangesMetaSince() which returns only metadata (actor, time, message, hash)
 * without materializing full document snapshots — this is fast (~1ms for hundreds
 * of changes).
 *
 * @param doc - The Automerge document to read history from
 * @param opts.limit - Maximum number of entries to return (default: all)
 * @param opts.sinceHeads - Only return changes after these heads (for pagination)
 * @returns Array of AuditEntry objects, newest first
 */
export function getAuditLog(
  doc: Doc,
  opts: { limit?: number; sinceHeads?: string[] } = {},
): AuditEntry[] {
  const sinceHeads = opts.sinceHeads ?? [];

  // getChangesMetaSince returns metadata for all changes after the given heads.
  // Passing [] gets all changes from the beginning.
  const metas: ChangeMeta[] = (Automerge as any).getChangesMetaSince(doc, sinceHeads);

  const entries: AuditEntry[] = [];

  for (const meta of metas) {
    if (!meta.message) continue;

    let parsed: ChangeMessage;
    try {
      parsed = JSON.parse(meta.message);
    } catch {
      // Skip non-JSON messages (e.g., migration changes)
      continue;
    }

    // Validate it has the expected structure
    if (!parsed.action || !parsed.target) continue;

    entries.push({
      action: parsed.action,
      actorId: parsed.actorId ?? "",
      actorName: parsed.actorName ?? parsed.actorId ?? "unknown",
      target: parsed.target,
      details: parsed.details ?? {},
      // Automerge stores time as Unix seconds; convert to milliseconds
      timestamp: meta.time * 1000,
      hash: meta.hash,
    });
  }

  // Return newest first
  entries.reverse();

  if (opts.limit !== undefined) {
    return entries.slice(0, opts.limit);
  }

  return entries;
}

/**
 * Get the total number of audit entries in the history.
 * Cheaper than getAuditLog() when you only need the count.
 */
export function getAuditLogCount(doc: Doc): number {
  const metas: ChangeMeta[] = (Automerge as any).getChangesMetaSince(doc, []);
  let count = 0;
  for (const meta of metas) {
    if (!meta.message) continue;
    try {
      const parsed = JSON.parse(meta.message);
      if (parsed.action && parsed.target) count++;
    } catch {
      continue;
    }
  }
  return count;
}

File: src/rust/crates/agt-lib/src/history.rs
//! Audit log reconstruction from Automerge's built-in change history.

use automerge::AutoCommit;

use crate::operations::ChangeMessage;
use crate::schema::AuditEntry;

/// Reconstruct the audit log from Automerge's change history.
/// Returns entries newest first.
pub fn get_audit_log(
    doc: &mut AutoCommit,
    limit: Option<usize>,
    _since_heads: Option<&[automerge::ChangeHash]>,
) -> Vec<AuditEntry> {
    let changes = doc.get_changes(&[]);
    let mut entries = Vec::new();

    for change in changes {
        let message = match change.message() {
            Some(msg) if !msg.is_empty() => msg,
            _ => continue,
        };

        let parsed: ChangeMessage = match serde_json::from_str(message) {
            Ok(p) => p,
            Err(_) => continue,
        };

        if parsed.action.is_empty() || parsed.target.is_empty() {
            continue;
        }

        entries.push(AuditEntry {
            action: parsed.action,
            actor_id: parsed.actor_id,
            actor_name: parsed.actor_name,
            target: parsed.target,
            details: parsed
                .details
                .unwrap_or(serde_json::Value::Object(Default::default())),
            timestamp: change.timestamp() * 1000,
            hash: change.hash().to_string(),
        });
    }

    entries.reverse();

    if let Some(limit) = limit {
        entries.truncate(limit);
    }

    entries
}

/// Get the total number of audit entries in the history.
pub fn get_audit_log_count(doc: &mut AutoCommit) -> usize {
    let changes = doc.get_changes(&[]);
    let mut count = 0;
    for change in changes {
        if let Some(msg) = change.message()
            && let Ok(parsed) = serde_json::from_str::<ChangeMessage>(msg)
            && !parsed.action.is_empty()
            && !parsed.target.is_empty()
        {
            count += 1;
        }
    }
    count
}
